indexer/SymbolFormatter.cc

```diff
@@ -40,44 +41,194 @@ std::string_view SymbolFormatter::getMacroSymbol(clang::SourceLocation defLoc) {
+          return fmt::format("{}$anontype_{}_{}#", contextSymbol,
+                             hashValue.rawValue, counter);
+        }
+      }
+      return fmt::format("{}$anontype_{}#", contextSymbol, counter);
+    }
+    llvm::raw_string_ostream os(this->scratchBuffer);
+    static_cast<const clang::NamedDecl *>(tagDecl)->printName(os);
+    std::string out{fmt::format("{}{}#", contextSymbol, this->scratchBuffer)};
+    this->scratchBuffer.clear();
+    return out;
+  });
+}
+
+std::optional<std::string_view> SymbolFormatter::getEnumConstantSymbol(
+    const clang::EnumConstantDecl *enumConstantDecl) {
+  return this->getSymbolCached(enumConstantDecl, [&]() -> std::string {
+    auto parentEnumDecl =
+        llvm::dyn_cast<clang::EnumDecl>(enumConstantDecl->getDeclContext());
+    ENFORCE(parentEnumDecl,
+            "decl context for EnumConstantDecl should be EnumDecl");
+    if (!parentEnumDecl) {
+      return {};
+    }
+    std::optional<std::string_view> optContextSymbol =
+        parentEnumDecl->isScoped()
+            ? this->getEnumSymbol(parentEnumDecl)
+            : this->getContextSymbol(parentEnumDecl->getDeclContext());
+    if (!optContextSymbol.has_value()) {
+      return {};
+    }
+    std::string out{fmt::format("{}{}.", optContextSymbol.value(),
+                                toStringView(enumConstantDecl->getName()))};
+    return out;
+  });
+}
+
+std::optional<std::string_view>
+SymbolFormatter::getEnumSymbol(const clang::EnumDecl *enumDecl) {
+  return this->getTagSymbol(static_cast<const clang::TagDecl *>(enumDecl));
+}
+
+std::optional<std::string_view>
+SymbolFormatter::getNamespaceSymbol(const clang::NamespaceDecl *namespaceDecl) {
+  auto symbol = this->getSymbolCached(namespaceDecl, [&]() -> std::string {
+    if (namespaceDecl->isAnonymousNamespace()) {
+      auto mainFileId = this->sourceManager.getMainFileID();
+      ENFORCE(mainFileId.isValid());
+      auto path = this->getCanonicalPath(mainFileId);
+      if (!path.has_value()) {
+        // Strictly speaking, this will be suboptimal in the following case:
+        // - header 1: in source tree (has canonical path), uses namespace {..}
+        // - header 2: in source tree (has canonical path), uses namespace {..}
+        // - generated C++ file: in build tree only (no canonical path), and
+        //   includes header 1 and 2.
+        // We will not emit a symbol that connects the anonymous namespace
+        // in header 1 and header 2. However, that is OK, because it is unclear
+        // how to handle this case anyways, and anonymous namespaces are
+        // rarely (if ever) used in headers.
+        return "";
+      }
+      fmt::format_to(std::back_inserter(this->scratchBuffer), "$ANON/{}\0",
+                     path->asStringView());
+    } else {
+      llvm::raw_string_ostream os(this->scratchBuffer);
+      namespaceDecl->printQualifiedName(os);
+      absl::StrReplaceAll({{"::", "/"}}, &this->scratchBuffer);
```

Comment:
Add comment here that the string replacement is justified because namespaces can only be present and not in types etc.